### ISSUE TITLE
Reaction runs too often?

### ISSUE THREAD

**Issue by liron00** [mobxjs/mobx#278]
```
class Foo {
  @observable x = 7

  constructor() {
    super()

    reaction(
      () => this.x >= 10,
      large => console.log('large', large)
    )

    setInterval(
      () => {
        this.x += 1
        console.log('set x to', this.x)
      },
      1000
    )
  }
}
```

Console output:

```
large false
set x to 8
large false
set x to 9
large true
set x to 10
```

Shouldn't it only run the side effect once until `large` changes to `true`?

**Comment by mweststrate** [comment]
no, you need `when` for that.

Op zo 29 mei 2016 05:22 schreef Liron Shapira <notifications@github.com>:

**Comment by liron00** [comment]
Hm I didn't really mean "once". What I mean is, if the computed expression gets evaluated again but still returns the same value as before, in this case false, is it really the intended behavior to call the side effect again with the same value?

**Comment by mweststrate** [comment]
Yep that is intended indeed (well kind of, unless the general feeling is that the semantics should be different :)).

You can achieve the desired behavior by using `computed(() => this.x > 10).observe(action(large => ...))`  (the `action` is optionally but idiomatic)

**Comment by liron00** [comment]
Well the docs for `reaction` say:

> Reaction is roughly speaking sugar for: `computed(expression).observe(action(sideEffect))` or `autorun(() => action(sideEffect)(expression)`

I think I prefer those semantics.

**Comment by mweststrate** [comment]
Thought about for a while, I think you are right. I think the nice thing of what you are proposing is that it forces people to create a pure excerpt of the data they want to react to. For example:

`reaction(() => { user.name; user.avatar; return user; }, person => {})`, will no longer react because the same user is returned over and over again. (Ok, probably nobody did this, just expecting the unexpected ;-))

Instead
`reaction(() => [user.name, user.avatar], ([name, avatar]) => {})` should be used which makes it a lot more explicit.

I will implement the semantics as suggested, and make sure `reaction(asStructure(() => ...), ...)`  is supported as well to demand explicit structural comparison of the return structure.

**Comment by liron00** [comment]
Sounds good to me. Btw in case you haven't seen this, here is Liebniz to your Newton: https://github.com/ds300/derivablejs

Check out the API, lots of interesting ideas. The biggest difference is that MobX's API has the primitive concept of a mutable object while in Derivable the only mutable thing is an Atom (but you can set the atom to have a slightly different immutableJS clone of the object it used to have).

Also here's my own crappy predecessor I was using before you guys came on the scene - https://github.com/quixey/jframework/

**Comment by mweststrate** [comment]
Released as part of 2.3
